src/core/polyhedral/codegen_llvm.cc

```diff
@@ -355,16 +355,34 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
   }
 };
 
+template <typename CI, typename CO, typename Pred>
+bool all(const CI& inputs, const CO& outputs, Pred p) {
+  for (const auto& v : inputs) {
+    if (not p(v)) {
+      return false;
+    }
+  }
+  for (const auto& v : outputs) {
+    if (not p(v)) {
+      return false;
+    }
+  }
+  return true;
+}
+
 class LLVMCodegen {
-  void collectTensor(const Halide::OutputImageParam& t) {
+  llvm::Type* convertTensorToType(const Halide::OutputImageParam& t) {
     auto sizes =
         getTensorSizesWithoutLeadingDim(t, scop_.globalParameterContext);
     if (not sizes.empty()) {
-      args_.emplace_back(
-          makePtrToArrayType(halide_cg.llvm_type_of(t.type()), sizes));
+      return makePtrToArrayType(halide_cg.llvm_type_of(t.type()), sizes);
     } else {
-      args_.emplace_back(halide_cg.llvm_type_of(t.type())->getPointerTo());
+      return halide_cg.llvm_type_of(t.type())->getPointerTo();
```

Comment:
just curious: what is `halide_cg`? also use `camelCase` :)

Reply:
@wsmoses  introduced it

Reply:
It translates from Halide to LLVM